File: LeetCode/121.Best_Time_to_Buy_and_Sell_Stock/Mario/121.Best_Time_to_Buy_and_Sell_Stock.java
class Solution {
    public int maxProfit(int[] prices) {
        // dp[i][0] sell
        // dp[i][1] buy/hold
        int n = prices.length;
        int[][] dp = new int[n][2];

        // init dp
        dp[0][0] = 0;
        dp[0][1] = -prices[0];

        // iterate
        for (int i = 1; i < n; i++) {
            dp[i][0] = Math.max(dp[i - 1][0], dp[i - 1][1] + prices[i]);
            dp[i][1] = Math.max(dp[i - 1][1], -prices[i]);
        }

        // for (int[] x : dp) {
        // System.out.print("\t" + x[0]);
        // }
        // System.out.println();
        // for (int[] x : dp) {
        // System.out.print("\t" + x[1]);
        // }

        return dp[n - 1][0];
    }
}

File: LeetCode/121.Best_Time_to_Buy_and_Sell_Stock/121.Best_Time_to_Buy_and_Sell_Stock.swift
class Solution {
    func maxProfit(_ prices: [Int]) -> Int {
        let n = prices.count
        var min_p = prices[0]
        var res = 0

        for i in 1 ..< n {
            res = max(prices[i] - min_p, res)
            min_p = min(min_p, prices[i])
        }

        return res
    }
}
